### PR TITLE
[MRG] implement a simple ZipFileLinearIndex class

### LINKED ISSUE #1320: distributing zip files of signatures, alone.
with greyhound #1226 #1238 and prefetch #1310, it would be really useful to be able to just distribute single zip files that contained signatures and nothing else.

ISTR @luizirber talking about this somewhere as distributing SBTs with no internal nodes, but I'm wondering if we even need JSON indices at all? I guess it wouldn't hurt, but I'm perfectly happy to not be required to load a big JSON at all.

### PR DESCRIPTION
**Pull request by ctb**
This PR implements linear indices that are just zipped collections of signatures. Fixes #1320.

Example usage:
```
zip -r test.zip podar-ref/*.fa.sig
sourmash compare test.zip
sourmash search podar-ref/1.fa.sig test.zip
```

A few notes -
* this is not an indexed database, in that search is linear..
* there are no internal sourmash commands for adding signatures or saving zip files; use `zip` at the command line.
* the file is tested to see if it's an SBT first.
* unlike indexed databases, the signatures in the zipfile can be incompatible ksizes/moltypes; selectors work as normal.
* in order to be loaded, the filename must end in .zip

I'm a little concerned at how easy this was, TBH... what am I missing? 😅 

Questions:
- do we want to support zipfile loading in directory traversal (e.g. `sourmash sig describe .`) and list-of-file inputs (`--from-file`)?

TODO:
- [x] add tests for command line usage
- [x] add tests for API usage, include force
- [x] update docs to describe new feature slash expand "collections of signatures" documentation vs "indexes"
- [ ] evaluate performance on e.g. all of GTDB
- [x] look at [DirectoryIndex issue](https://github.com/dib-lab/sourmash/issues/810) re caching and/or expansion of Index API to support "this sig was loaded from this source, specifically" - NOTE, done over in #1392 currently.

### CODE REVIEW

**Comment by ctb** [comment]
Interested in comments, @bluegenes @luizirber! I still have a few tests to write and some documentation to update, but curious what you think.

So far I really like the convenience of supporting everything in one file, and it's particularly nice that it supports incompatible signatures (unlike the indexed data types, for now).

**Comment by bluegenes** [comment]
love this -- much easier to distribute one file than many, plus compression! A couple thoughts:

-  `sourmash sig describe` --> yes, would be nice to support usage, but output may not be as useful? Likely too many sigs for full output, and `sourmash sig describe <zipfile> | head` would no longer represent the parameters of all sigfiles, since it's possible to incorporate incompatible signatures. If enabled, maybe we could summarize the sig params (e.g. num sigs per ksize/alpha/scaled ) as tsv or tabbed terminal output? Or have a quick default option that gives info for first sig, and `summarize` option that gives full summary? Somewhat related to #1190.

- loading via `--from-file` --> yes, please!

- Future thoughts: Are you set on zipfile generation being external? If not --
  - Zipped output from `search` or `gather`(when outputting matches/unmatched) could be useful?  
  - Would we want to enable building a `sigs.zip` file from a file with a list of sigs? 
  - If sketching a bunch of sigs, could we write directly to a zip file? Not sure I would actually do this for non `--singleton` cases as it would eliminate parallelization, but just throwing it out there.

**Comment by ctb** [comment]
Love the comments, excellent brainstorming :). I will probably punt most of them to new issues as out of my desired scope for this PR, but I think they're great next steps!

Note that outputting '.gz' formats is just as good as a zip file when compression is desired; zip files are mostly nice for situations where you have a directory of things to distribute, and want to be able to update them etc (the random access case).

I like the idea of a `sourmash sig summary` command for collections, FWIW.

**Comment by bluegenes** [comment]
>Note that outputting '.gz' formats is just as good as a zip file when compression is desired; zip files are mostly nice for situations where you have a directory of things to distribute, and want to be able to update them etc (the random access case).

Can you expand on this a bit? I guess you mean that for compression, we could gzip the individual sigfiles, and then optionally zip the directory, which enables distribution/updating/etc. But I guess I haven't been thinking about compression much-- do we currently enable reading/writing `gz` sigfiles? Or is it just that `gz` is easier to enable for sig output? How do the pros (space) / cons (loading time??) shake out? Sry, naiive compression q's -- happy to drop this in a different issue if better :).

**Comment by luizirber** [comment]
On the plus side, this is a simple implementation, with easy-to-use commands in the CLI, and covers an increasingly common use case.

But =]
- It's putting together `Index` and `Storage`. Nothing necessarily wrong with that, but the benefit of going thru the `Storage` layer is being able to share it with other indices (especially SBTs at the moment).
- This is an additional file format to support (on top of signatures, LCA, SBT, and probably file-of-files should be considered a format too). And the file generation is "outside" our control, so it's harder to version.
- In #1201 I was adding a new version of the SBT (`v7`, I think?) that splits internal nodes and leaves/signatures. This could be used here (where `signatures/` would be a dir inside the zip file with the signatures), and make this and the SBT zip formats compatible.
- We already do quite a bit of adhoc checks to support Zip/JSON SBTs, and this PR adds another one (distinguish between SBT and Linear zip files)
- There is a very old (and unfinished) PR for saving full sigs into the SBT #366. There was no `Index` or selectors at that point, but would be similar to this use case.
 
> Questions:
> * are there any performance concerns with the way I'm doing this?

Not at first sight, but we will only know when we measure =]

> do we want to support loading .sig.gz files from within zip files?

This already works transparently, because `niffler` opens the data (compressed or uncompressed) on the Rust side. No?

----

So, I'm +1 on this PR. I think it fragments a bit the codebase and makes it harder to maintain in the long run, but it is useful NOW, which is more important =]

**Comment by ctb** [comment]
some responses and thoughts -

>>Note that outputting '.gz' formats is just as good as a zip file when compression is desired; zip files are mostly nice for situations where you have a directory of things to distribute, and want to be able to update them etc (the random access case).

@bluegenes:
>Can you expand on this a bit? I guess you mean that for compression, we could gzip the individual sigfiles, and then optionally zip the directory, which enables distribution/updating/etc. But I guess I haven't been thinking about compression much-- do we currently enable reading/writing `gz` sigfiles? Or is it just that `gz` is easier to enable for sig output? How do the pros (space) / cons (loading time??) shake out? Sry, naiive compression q's -- happy to drop this in a different issue if better :).

sourmash can load multiple signatures from `.sig.gz` files just fine. So we could output matches etc in `.sig.gz` format it all we wanted was compression.

So the question becomes, what does a .zip file bring that goes beyond compression?

The main advantages that I see -
* the ability to list and manipulate the file contents from outside sourmash using a variety of tools
* storing non `.sig` files, e.g. files that contain taxonomy info per https://github.com/dib-lab/sourmash/issues/969#issuecomment-670249777, or maybe manifests (the way SBTs have the `.sbt.json` "directory" files)
* faithfully representing directory structures and sensible filenames (which can be good or bad, but it's nice to have the option)
* a fairly good intuitive explanation for "what is this file?" - it's merely a list of signature files that you put in a zip file.

That is, zip files are a package of compressed files and directories, not a compressed package of signatures.

@luizirber:
> - It's putting together `Index` and `Storage`. Nothing necessarily wrong with that, but the benefit of going thru the `Storage` layer is being able to share it with other indices (especially SBTs at the moment).

This is an excellent point, and one I want to dig into. Are you think maybe that `LinearIndex` could be stored in a `ZipStorage`, or an `lca.json` file could be stored on IPFS, potentially?

> - This is an additional file format to support (on top of signatures, LCA, SBT, and probably file-of-files should be considered a format too). And the file generation is "outside" our control, so it's harder to version.
> - We already do quite a bit of adhoc checks to support Zip/JSON SBTs, and this PR adds another one (distinguish between SBT and Linear zip files)

I'm not too concerned about the versioning, since zipfile format is pretty stable, is supported by Python, and presumably is supported by Rust.

But your point about this being Yet Another file format to support is well taken. How many more `if` statements do we want in `sourmash_args`? :laughing: 

> - In #1201 I was adding a new version of the SBT (`v7`, I think?) that splits internal nodes and leaves/signatures. This could be used here (where `signatures/` would be a dir inside the zip file with the signatures), and make this and the SBT zip formats compatible.

So this brings up a key question: do we want to require that there be a manifest? A required manifest can add loading overhead and maintenance overhead (if we update the things in the zip file, we need to update the manifest).

In the greyhound use case we simply want to be able to iterate across the signatures as quickly as possible, and I think the current SBT format is ill suited for that.

> - There is a very old (and unfinished) PR for saving full sigs into the SBT #366. There was no `Index` or selectors at that point, but would be similar to this use case.

k, I'll take a look.

> > do we want to support loading .sig.gz files from within zip files?
> 
> This already works transparently, because `niffler` opens the data (compressed or uncompressed) on the Rust side. No?

Yes, it should, but we need to recognize `.sig.gz` files in the internal logic of the Zipfile loading.
 
> So, I'm +1 on this PR. I think it fragments a bit the codebase and makes it harder to maintain in the long run, but it is useful NOW, which is more important =]

This matches my hot take (once I'd finished getting it working) - good functionality, but not that cleanly integrated into sourmash. There's no immediate hurry, so let's keep talking.

So a follow-on question:

What about having some logic where the Storage tells us what kind of file it is? Then we can implement ZipStorage logic to tell us if it's an SBT or something else.

**Comment by ctb** [comment]
Wrote a bit about [how a `DirectoryIndex` might work in terms of supporting signature origin tracking](https://github.com/dib-lab/sourmash/issues/810#issuecomment-787470143), because I'm wondering if we could expand the `Storage` API in `sbt_storage.py` to support a wider range of retrieval approaches (and then maybe build `DirectoryIndex `on top of `ZipStorage` as a replacement for this `ZipFileLinearIndex` class).

tl;dr I am wondering if we should provide a URL designator for signature origins?

 In doing so, I came across [this interesting discussion about how to write URLs for files in zip archives](https://github.com/frictionlessdata/specs/issues/137) (from [frictionlessdata](https://frictionlessdata.io/) that does a nice breakdown of some issues with origin URLs. 

Which led me to [this issue on bdbags](https://github.com/dib-lab/sourmash/issues/991) 🤣 so it's all getting very complicated!

**Comment by ctb** [comment]
while I'm weaving a web of issues this morning...

* when thinking about [manifests #1352](https://github.com/dib-lab/sourmash/issues/1352), a nice clear use case for zip files emerged: the manifest can be part of the collection.
* would be interesting to trial-implement a SQL storage (perhaps in sqlite?) for signature collections.

**Comment by ctb** [comment]
further brain breaking fun thoughts:

aren't indexes separate from collections? indexes are indices ON collections of sketches...

* collections support dumb traversal, along with underlying signature/sketch storage and retrieval
* `Index` objects provide fast queries of various types (search, gather)

thinking down this road a bit further,

* collections would support specific storage types (file system, zip, ipfs, redis, sql) - we could provide a URL schema that is externally resolvable, at least by other commands within sourmash
* manifests would be information about collections, optionally provided
* indexes would be separate [mixins](https://en.wikipedia.org/wiki/Mixin) that support better querying

there is a slightly messy interplay with sketch types and selectors and tagging -

* indexes are generally going to be on a particular subset of compatible sketches within a collection;
* so a search using an index would be able to look at a subset of your collection;
* and you might want to refine the search results by selecting on tags, or taxonomy, or something, AFTER the search is done;
* you run into the (far distant future) problem that there maybe an index on some subset of a collection, and you want to use it for searching those, but not for the things not included in the index;

but this all seems like buying ourselves conceptual trouble for the future that might not be that interesting in practice...

note we reframe `sourmash index` as generating an index on a collection, e.g. an LCA/revindex index on a collection of signatures. Of course some of the indices (LCA currently) store the signatures internally, others don't need to but may (SBTs).

This further opens up the idea of commands to generate manifests on collections separately from indices.

**Comment by ctb** [comment]
Here's some simple code to load a pile o' signatures into a zipfile:

```python
#! /usr/bin/env python
import sys
import zipfile
import sourmash
import argparse


def main():
    p = argparse.ArgumentParser()
    p.add_argument('zipfile')
    p.add_argument('signatures', nargs='+')
    args = p.parse_args()

    zf = zipfile.ZipFile(args.zipfile, 'w', zipfile.ZIP_DEFLATED)

    n = 0
    for filename in args.signatures:
        print(f"reading signatures from '{filename}'")
        for sig in sourmash.load_file_as_signatures(filename):
            md5 = 'signatures/' + sig.md5sum() + '.sig'
            sigstr = sourmash.save_signatures([sig])
            zf.writestr(md5, sigstr)
            n += 1

    print(f"wrote {n} signatures to '{args.zipfile}'")

    return 0


if __name__ == '__main__':
    sys.exit(main())
```

**Comment by ctb** [comment]
more idle thinking, would the results of `prefetch` be ...another index? or is that too query-specific? hmm. probably. But they could be a collection.

...perhaps with a new kind of index, a summary sketch that contains aggregate hashes plus their abundances.

or maybe that's just an lca/revindex index.

which leads me to think that the output of commands like prefetch and `sourmash gather --save-matches` could usefully be an index like LCA/revindex instead of a zipfile - outputting an index no longer restricts downstream usage, since all of those outputs will be treated the same by sourmash signature search/manipulation commands.

**Comment by bluegenes** [comment]
Could the result of `prefetch` be an alternate manifest or a list of names that can be used to select signatures to load from the existing collection/index? 

Would be really nice to avoid saving a `matches.sig` intermediate file, if it's not too hard to just load the relevant sigs from a collection.

**Comment by ctb** [comment]
> Could the result of `prefetch` be an alternate manifest or a list of names that can be used to select signatures to load from the existing collection/index?
> 
> Would be really nice to avoid saving a `matches.sig` intermediate file, if it's not too hard to just load the relevant sigs from a collection.

I like this - I wonder if we could support multiple output formats based on output filename extension, i.e. if you provide a .zip extension on the `--save-matches` filename it goes to a .zip file, or a .lca.json extension, etc. Then if you added a picklist or manifest extension it would save appropriately.

**Comment by ctb** [comment]
With #1406 and #1420 being merged, loading of databases and selection of compatible signatures has improved muchly; while there are still many to be improved, I think this PR is reasonably well scoped and fully baked.

I'll take care of creating new issues based on the conversations above before merge, but for now I am happy to say: ready for review! pls review at your leisure, @bluegenes! (because I know you're enthusiastic about this feature :)

**Comment by ctb** [comment]
(the test failures appear to be related to something else going on with the docs; I'll look into this separately)

File: doc/command-line.md [hunk 1]
```diff
@@ -842,7 +842,9 @@ scaled values will be made compatible.
 ### Storing (and searching) signatures
   
 Backing up a little, there are many ways to store and search
-signatures.
+signatures. sourmash supports storing and loading signatures from JSON
```
Comment (bluegenes) [inline doc/command-line.md]:
```suggestion
signatures. `sourmash` supports storing and loading signatures from JSON
```

File: tests/test_index.py [hunk 2]
```diff
@@ -401,6 +401,207 @@ def test_index_same_md5sum_zipstorage(c):
...
+    # test command-line search/gather of combined zipfile with dayhoff sigs
+    sigfile1 = utils.get_test_data('prot/dayhoff/GCA_001593925.1_ASM159392v1_protein.faa.gz.sig')
+    db_out = utils.get_test_data('prot/all.zip')
+
+    c.run_sourmash('search', sigfile1, db_out, '--threshold', '0.0')
+    assert '2 matches:' in c.last_result.out
+
+    c.run_sourmash('gather', sigfile1, db_out, '--threshold', '0.0')
+    assert 'found 1 matches total' in c.last_result.out
+    assert 'the recovered matches hit 100.0% of the query' in c.last_result.out
```
Comment (bluegenes) [inline tests/test_index.py:485]: do we need a test for running search/gather on a zipfile index with an incompatible sig?

File: tests/test_index.py [hunk 3]
```diff
@@ -676,3 +877,17 @@ def test_multi_index_load_from_pathlist_2(c):
 
     with pytest.raises(ValueError):
         mi = MultiIndex.load_from_pathlist(file_list)
+
+
+@utils.in_tempdir
+def test_multi_index_load_from_pathlist_3_zipfile(c):
+    # can we load zipfiles in a pathlist? yes please.
```
Comment (bluegenes) [inline tests/test_index.py:897]: yay :)

File: src/sourmash/index.py [hunk 4]
```diff
@@ -219,7 +220,66 @@ def select(self, **kwargs):
...
+                fp = self.zf.open(zipinfo)
+
+                # now load all the signatures and select on ksize/moltype:
+                selection_dict = self.selection_dict
+                for ss in load_signatures(fp):
+                    if selection_dict:
+                        if select_signature(ss, **self.selection_dict):
+                            yield ss
+                    else:
+                        yield ss
```
Comment (bluegenes) [inline src/sourmash/index.py:276]: trying to wrap my head around `traverse_yield_all` --

do all files get loaded into generator, but non-sig files are never returned bc they can't be loaded as sigs (and thus won't be selected)?

File: src/sourmash/index.py [hunk 5]
```diff
@@ -219,7 +220,66 @@ def select(self, **kwargs):
...
+        for zipinfo in self.zf.infolist():
+            # should we load this file? if it ends in .sig OR we are forcing:
+            if zipinfo.filename.endswith('.sig') or \
+               zipinfo.filename.endswith('.sig.gz') or \
+               self.traverse_yield_all:
+                fp = self.zf.open(zipinfo)
+
+                # now load all the signatures and select on ksize/moltype:
+                selection_dict = self.selection_dict
+                for ss in load_signatures(fp):
```
Comment (bluegenes) [inline src/sourmash/index.py:271]: Why `load_signatures` rather than `load_file_as_signatures`, which would enable loading index files that are nested in the zipfile? Do we want to avoid potentially complicated nesting?

**Review by bluegenes — COMMENTED** [review]
looking good to me, just a few questions /comments!

File: src/sourmash/index.py (hunk 5 above)
Comment (ctb) [reply src/sourmash/index.py:271]: good question! `load_file_as_signatures` takes a filename, while `load_signatures` can take a filename, a string, or a file handle - which is what it's doing here.

File: src/sourmash/index.py (hunk 4 above)
Comment (ctb) [reply src/sourmash/index.py:276]: correct!

Comment (ctb) [reply src/sourmash/index.py:276]: although, looking at it, I'm a little surprised this works. Will dig.

File: tests/test_index.py (hunk 2 above)
Comment (ctb) [reply tests/test_index.py:485]: added in 6ce3ce57 - conveniently, this is all handled by the new `select` approach, but worth testing explicitly!

File: src/sourmash/index.py (hunk 5 above)
Comment (bluegenes) [reply src/sourmash/index.py:271]: ok! Guess I'm a little confused on deprecation of `load_signatures`, then --what is the plan for this functionality in the future? Or are we just deprecating from the public api?

**Review by bluegenes — APPROVED** [review]

File: src/sourmash/index.py (hunk 4 above)
Comment (ctb) [reply src/sourmash/index.py:276]: ok, commented in 56184ffc - `load_signatures` will fail silently on invalid JSON/files that aren't loadable, and return nothing. (If you pass in `do_raise=True` it will raise a `ValueError`)

File: src/sourmash/index.py (hunk 5 above)
Comment (ctb) [reply src/sourmash/index.py:271]: :) good question. Yes, deprecating from the public API, and also --

* maybe changing `load_file_as_signatures` to support file handles https://github.com/dib-lab/sourmash/issues/1312
* breaking up `load_signatures` to be less complicated, implementation-wise https://github.com/dib-lab/sourmash/issues/1060

Comment (bluegenes) [reply src/sourmash/index.py:271]: thx!

**Comment by bluegenes** [comment]
> Here's some simple code to load a pile o' signatures into a zipfile:
> 
> ```python
> #! /usr/bin/env python
> import sys
> import zipfile
> import sourmash
> import argparse
> 
> 
> def main():
>     p = argparse.ArgumentParser()
>     p.add_argument('zipfile')
>     p.add_argument('signatures', nargs='+')
>     args = p.parse_args()
> 
>     zf = zipfile.ZipFile(args.zipfile, 'w')
> 
>     n = 0
>     for filename in args.signatures:
>         print(f"reading signatures from '{filename}'")
>         for sig in sourmash.load_file_as_signatures(filename):
>             md5 = 'signatures/' + sig.md5sum() + '.sig'
>             sigstr = sourmash.save_signatures([sig])
>             zf.writestr(md5, sigstr)
>             n += 1
> 
>     print(f"wrote {n} signatures to '{args.zipfile}'")
> 
>     return 0
> 
> 
> if __name__ == '__main__':
>     sys.exit(main())
> ```

FYI this may have issues with duplicated md5sums, see https://github.com/dib-lab/sourmash/issues/1483#issuecomment-827107220
